crates/ra_ide/src/hover.rs

```diff
@@ -335,6 +346,277 @@ fn hover_for_definition(db: &RootDatabase, def: Definition) -> Option<Markup> {
+    out
+}
+
+#[derive(PartialEq, Eq, Hash, Copy, Clone, Debug)]
+enum Namespace {
+    Types,
+    Values,
+    Macros,
+}
+
+lazy_static!(
+    /// Map of namespaces to identifying prefixes and suffixes as defined by RFC1946.
+    static ref NS_MAP: HashMap<Namespace, (HashSet<&'static str>, HashSet<&'static str>)> = hashmap!{
+        Namespace::Types => (hashset!{"type", "struct", "enum", "mod", "trait", "union", "module"}, hashset!{}),
+        Namespace::Values => (hashset!{"value", "function", "fn", "method", "const", "static", "mod", "module"}, hashset!{"()"}),
+        Namespace::Macros => (hashset!{"macro"}, hashset!{"!"})
+    };
+);
+
+impl Namespace {
+    /// Extract the specified namespace from an intra-doc-link if one exists.
+    fn from_intra_spec(s: &str) -> Option<Self> {
+        NS_MAP
+            .iter()
+            .filter(|(_ns, (prefixes, suffixes))| {
+                prefixes
+                    .iter()
+                    .map(|prefix| {
+                        s.starts_with(prefix)
+                            && s.chars()
+                                .nth(prefix.len() + 1)
+                                .map(|c| c == '@' || c == ' ')
+                                .unwrap_or(false)
+                    })
+                    .any(|cond| cond)
+                    || suffixes
+                        .iter()
+                        .map(|suffix| {
+                            s.starts_with(suffix)
+                                && s.chars()
+                                    .nth(suffix.len() + 1)
+                                    .map(|c| c == '@' || c == ' ')
+                                    .unwrap_or(false)
+                        })
+                        .any(|cond| cond)
+            })
+            .map(|(ns, (_, _))| *ns)
+            .next()
+    }
+}
+
+// Strip prefixes, suffixes, and inline code marks from the given string.
+fn strip_prefixes_suffixes(mut s: &str) -> &str {
+    s = s.trim_matches('`');
+    NS_MAP.iter().for_each(|(_, (prefixes, suffixes))| {
+        prefixes.iter().for_each(|prefix| s = s.trim_start_matches(prefix));
+        suffixes.iter().for_each(|suffix| s = s.trim_end_matches(suffix));
+    });
+    s.trim_start_matches("@").trim()
+}
+
+/// Try to resolve path to local documentation via intra-doc-links (i.e. `super::gateway::Shard`).
+///
+/// See [RFC1946](https://github.com/rust-lang/rfcs/blob/master/text/1946-intra-rustdoc-links.md).
+fn try_resolve_intra(
+    db: &RootDatabase,
+    definition: &Definition,
+    link_text: &str,
+    link_target: &str,
+) -> Option<(String, String)> {
+    // Set link_target for implied shortlinks
+    let link_target =
+        if link_target.is_empty() { link_text.trim_matches('`') } else { link_target };
+
+    // Namespace disambiguation
+    let namespace = Namespace::from_intra_spec(link_target);
+
+    // Strip prefixes/suffixes
+    let link_target = strip_prefixes_suffixes(link_target);
```

Comment:
Seems like this could use some types?

```
struct DocLink {
    path: &str,
    namespace: Option<Namespace>,
}

impl FromStr for DocLink { ... }
```

Reply:
I've debated this for a little bit and I'm not sure I think it's worth the mental overhead of another type for the increased type safety.

More than happy to defer to your judgement here if you still think it's a good idea.

Reply:
I always err on the side of adding types, on the grounds that *removing* type later is easier than adding it.